### PR TITLE
Fake clientset propagates namespace to objects on create/update

### LINKED ISSUE #48: Fake Clientset requires namespace be set on Create and Update but the real Clientset and API server do not
The fake `Clientset` requires creates and updates of namespaced objects to have the namespace specified in the target runtime object metadata while the same does not appear to be true for the real `Clientset` and API server.

In particular, something like this:
```go
func createConfigMap(c kubernetes.Interface) error {
	_, err := c.Core().ConfigMaps("namespace").Create(&v1.ConfigMap{
		ObjectMeta: v1.ObjectMeta{Name: "name"},
		Data:       map[string]string{"key": "value"},
	})
	return err
}
```
works against a real cluster with a real `Clientset` but fails with `no namespace specified for a namespaced kind /v1, Kind=ConfigMap` when using the fake.

---

It looks like Create and Update end up adding the object to the `tracker` in [`fixture.go`][1] where the [namespace is checked][2]. The current implementation passes the namespace from the target object without accounting for the namespace from the request.

In the real API server, namespaces are propagated from the request path to the target object when a namespace isn't set on the object. This happens when [`ValidNamespace`][3] is called out of [`BeforeCreate`][4] and [`BeforeUpdate`][5].

Should the behavior of the tracker be updated to propagate the namespace from the request to the object when it's missing on the object?

[1]: https://github.com/kubernetes/client-go/blob/5d8c36c93cf544e3bde13caa4c6222b65753ec3c/testing/fixture.go#L245-L251
[2]: https://github.com/kubernetes/client-go/blob/5d8c36c93cf544e3bde13caa4c6222b65753ec3c/testing/fixture.go#L289
[3]: https://github.com/kubernetes/kubernetes/blob/35a6bfbceed9d8bc779fbfe00963b7674ea68b77/pkg/api/context.go#L104-L110
[4]: https://github.com/kubernetes/kubernetes/blob/35a6bfbceed9d8bc779fbfe00963b7674ea68b77/pkg/api/rest/create.go#L63-L65
[5]: https://github.com/kubernetes/kubernetes/blob/35a6bfbceed9d8bc779fbfe00963b7674ea68b77/pkg/api/rest/update.go#L83-L85

### PR DESCRIPTION
**Pull request by sykesm**
**What this PR does / why we need it**:

Unlike the real Clientset, the fake Clientset requires creates and updates of namespaced objects to have the namespace specified in the target runtime object metadata. This difference forces API clients using the fake Clientset for test to propagate the namespaces to the runtime objects in the production code. This propagation is unnecessary and should be handled by the fake implementation.

**Which issue this PR fixes**: fixes kubernetes/client-go#48

**Special notes for your reviewer**:

**Release note**:
NONE

### CODE REVIEW

**Comment by sykesm** [comment]
Updated the commit with my corporate e-mail address and verified that I am covered by the existing IBM CLA.

**Comment by smarterclayton** [comment]
@k8s-bot ok to test

File: pkg/client/testing/core/fixture.go [hunk 1]
```diff
@@ -286,6 +297,15 @@ func (t *tracker) add(obj runtime.Object, replaceExisting bool) error {
 			return err
 		}
 
+		// Propagate namespace to the new object if hasn't already been set.
+		if len(newMeta.GetNamespace()) == 0 {
+			newMeta.SetNamespace(ns)
+		}
+
+		if newMeta.GetNamespace() != ns {
```
Comment (smarterclayton) [inline pkg/client/testing/core/fixture.go]: This is kind of awkward for what the tracker is supposed to do.  There are many tests that could span namespaces, so erroring out here seems incorrect.

**Comment by smarterclayton** [comment]
@soltysh because I know you were in this code recently

File: pkg/client/testing/core/fixture.go (hunk 1 above)
Comment (soltysh) [inline pkg/client/testing/core/fixture.go]: @smarterclayton but this mimics the behavior you have when invoking similar operations against regular cluster. Operations spanning multiple namespaces does not include `Create` or `Update`, these are always targeted against specific namespace, and this is what `add` doe here. I'd just update this to include the namespace names, like this:
```
the namespace from the provided object <newMeta.GetNamespace()> does not match the namespace <ns>
```

**Comment by soltysh** [comment]
@k8s-bot node e2e test this

**Comment by soltysh** [comment]
@smarterclayton this LGTM, if there are no objections I'd like to tag it for merge.

File: pkg/client/testing/core/fixture.go (hunk 1 above)
Comment (sykesm) [inline pkg/client/testing/core/fixture.go]: @soltysh added a commit to populate the error with information about the mismatched namespaces. Thanks for reviewing.

**Comment by soltysh** [comment]
@sykesm please squash your changes and it'll be good to merge.

**Comment by sykesm** [comment]
@soltysh Squashed. Thanks.

**Pull request merged**